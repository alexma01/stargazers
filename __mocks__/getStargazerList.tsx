import {Stargazer} from '../src/models/Stargazer';

export const stargazersMocks: Stargazer[] = [
  {
    username: 'test1',
    avatar: 'https://avatars.githubusercontent.com/u/18049606?v=4',
  },
  {
    username: 'test2',
    avatar: 'https://avatars.githubusercontent.com/u/18049606?v=4',
  },
  {
    username: 'test3',
    avatar: 'https://avatars.githubusercontent.com/u/18049606?v=4',
  },
  {
    username: 'test4',
    avatar: '',
  },
];
