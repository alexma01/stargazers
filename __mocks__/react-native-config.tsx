export default {
  GITHUB_TOKEN: '',
  GITHUB_ENDPOINT: 'https://api.github.com/',
};
