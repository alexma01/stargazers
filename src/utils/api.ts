import {Stargazer} from '../models/Stargazer';
import Config from 'react-native-config';

export const fetchStargazers = async (
  owner: string,
  repository: string,
  page: number,
  perPage: number,
): Promise<Stargazer[]> => {
  try {
    const url: string = `${Config.GITHUB_ENDPOINT}repos/${owner}/${repository}/stargazers?per_page=${perPage}&page=${page}`;
    const response: Response = await fetch(url, {
      headers: {
        Authorization: `Bearer ${Config.GITHUB_TOKEN}`,
      },
    });

    if (response.status === 200) {
      const datas = await response.json();
      const stargazers: Stargazer[] = datas.map((row: any) => {
        return {
          username: row.login,
          avatar: row.avatar_url,
        };
      });
      return stargazers;
    }
  } catch (error) {
    console.error(error);
  }
  return [];
};
