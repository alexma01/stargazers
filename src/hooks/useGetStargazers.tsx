import {useCallback, useRef, useState} from 'react';
import {Stargazer} from '../models/Stargazer';
import {fetchStargazers} from '../utils/api';

const useGetStargazers = () => {
  const prevOwner = useRef<string>();
  const prevRepository = useRef<string>();
  const prevPage = useRef<number>();
  const [data, setdata] = useState<Stargazer[]>([]);
  const [loading, setloading] = useState(false);
  const [error, seterror] = useState('');

  const fetchData = useCallback(
    async (
      ownerR: string,
      pageR: number,
      per_pageR: number,
      repositoryR: string,
    ) => {
      setloading(true);
      const stargazers = await fetchStargazers(
        ownerR,
        repositoryR,
        pageR,
        per_pageR,
      );
      console.log('FETCH', ownerR);
      console.log('FETCH', repositoryR);
      console.log('FETCH', pageR);
      console.log('FETCH', per_pageR);
      if (
        prevRepository.current === repositoryR &&
        prevOwner.current === ownerR &&
        prevPage.current !== pageR
      ) {
        setdata(stargazersPrev => [...stargazersPrev, ...stargazers]);
        prevPage.current = pageR;
        prevOwner.current = ownerR;
        prevRepository.current = repositoryR;
      } else {
        setdata(stargazers);
        prevPage.current = pageR;
        prevOwner.current = ownerR;
        prevRepository.current = repositoryR;
      }

      if (stargazers.length === 0) {
        seterror('NOT FOUND');
        setdata([]);
      }

      setloading(false);
    },
    [],
  );

  const onFetch = useCallback(
    (owner: string, page: number, per_page: number, repository: string) => {
      seterror('');
      if (owner && repository) {
        fetchData(owner, page, per_page, repository);
      }
    },
    [fetchData],
  );

  return {data, loading, error, onFetch};
};

export default useGetStargazers;
