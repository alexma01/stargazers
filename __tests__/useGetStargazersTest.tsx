import {renderHook, act} from '@testing-library/react-hooks';
import useGetStargazers from '../src/hooks/useGetStargazers';
import {stargazersMocks} from '../__mocks__/getStargazerList';
import * as api from '../src/utils/api';

describe('useGetStargazers', () => {
  it('should match the snapshot', async () => {
    jest.spyOn(api, 'fetchStargazers').mockImplementationOnce(() => {
      return Promise.resolve(stargazersMocks);
    });

    const {result} = renderHook(() => useGetStargazers());

    expect(result.current.loading).toBe(false);
    await act(async () => {
      result.current.onFetch('dsdsd', 1, 20, 'dsds');
    });
    expect(result.current.loading).toBe(false);
    expect(result.current.data).toMatchSnapshot();
  });

  it('should return empty list', async () => {
    const {result} = renderHook(() => useGetStargazers());
    expect(result.current.loading).toBe(false);
    await act(async () => {
      result.current.onFetch('', 1, 20, '');
    });
    expect(result.current.loading).toBe(false);
    expect(result.current.data).toMatchSnapshot();
  });
});
