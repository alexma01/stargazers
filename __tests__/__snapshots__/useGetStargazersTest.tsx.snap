// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useGetStargazers should match the snapshot 1`] = `
[
  {
    "avatar": "https://avatars.githubusercontent.com/u/18049606?v=4",
    "username": "test1",
  },
  {
    "avatar": "https://avatars.githubusercontent.com/u/18049606?v=4",
    "username": "test2",
  },
  {
    "avatar": "https://avatars.githubusercontent.com/u/18049606?v=4",
    "username": "test3",
  },
  {
    "avatar": "",
    "username": "test4",
  },
]
`;

exports[`useGetStargazers should return empty list 1`] = `[]`;
